esphome/components/voice_kit/__init__.py

```diff
@@ -1,31 +1,125 @@
+import hashlib
+from pathlib import Path
 import esphome.codegen as cg
 import esphome.config_validation as cv
 from esphome import pins
+from esphome import automation, core, external_files
 from esphome.components import i2c
-from esphome.const import CONF_ID, CONF_RESET_PIN
+from esphome.const import (
+    CONF_ID,
+    CONF_RAW_DATA_ID,
+    CONF_RESET_PIN,
+    CONF_URL,
+    CONF_VERSION,
+)
+from esphome.core import HexInt
 
 CODEOWNERS = ["@kbx81"]
 DEPENDENCIES = ["i2c"]
 
+CONF_FIRMWARE = "firmware"
+CONF_MD5 = "md5"
+
+DOMAIN = "voice_kit"
+
 voice_kit_ns = cg.esphome_ns.namespace("voice_kit")
 VoiceKit = voice_kit_ns.class_("VoiceKit", cg.Component, i2c.I2CDevice)
+VoiceKitFlashAction = voice_kit_ns.class_("VoiceKitFlashAction", automation.Action)
+
+
+def _compute_local_file_path(url: str) -> Path:
+    h = hashlib.new("sha256")
+    h.update(url.encode())
+    key = h.hexdigest()[:8]
+    base_dir = external_files.compute_local_file_dir(DOMAIN)
+    return base_dir / key
+
+
+def download_firmware(config):
+    url = config[CONF_URL]
+    path = _compute_local_file_path(url)
+    external_files.download_content(url, path)
+
+    try:
+        with open(path, "r+b") as f:
+            firmware_bin = f.read()
+    except Exception as e:
+        raise core.EsphomeError(f"Could not open firmware file {path}: {e}")
```

Comment:
Should be an Invalid raised here, and also try to avoid `except Exception` as it is too broad.
```suggestion
    except FileNotFoundError as e:
        raise cv.Invalid(f"Could not open firmware file {path}: {e}") from e
```